### DISCUSSION TITLE
Multipart/mixed content-type support

### DISCUSSION THREAD

**Discussion by raif-s-naffah** [rwf2/Rocket#2811] [Questions] [answered]
hi there,

i'm trying to implement some IEEE standard REST API that mandates handling `PUT` and `POST` requests  w/ [`multipart/mixed`][1] content type.

my tests w/ Rocket 5.1 so far indicate that using a Form like this one...
```rust
use rocket::FromForm;

#[derive(Debug, FromForm)]
struct OneStatement<'r> {
    doc: &'r str,
    files: Option<Vec<TempFile<'r>>>,
}
```
in a handler such as ...
```rust
#[put("/", data = "<form>")]
async fn put(mut form: Form<OneStatement<'_>>) -> MyResponse {}
```
fail w/ `415 (Unsupported Media Type)` presumably b/c Rocket doesn't recognize the `multipart/mixed` content-type header and consequently doesn't route the Request to the handler.

**q1**/ is there already some Rocket feature that makes this possible?

one alternative would be to handle the decoding and partitioning of the incoming data myself w/ the [Streaming][2] capability.

**q2**/ is this the right approach? and if it is...

**q3**/ _--although this is not specifically related to Rocket--_ can you recommend any 3rd-party crate(s) that do this **and** work well w/ Rocket?

thanks in advance + cheers;
rsn

[1]: <https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html>
[2]: <https://rocket.rs/guide/v0.5/requests/#streaming>

**Comment by the10thWiz** [comment]
> fail w/ 415 (Unsupported Media Type) presumably b/c Rocket doesn't recognize the multipart/mixed content-type header and consequently doesn't route the Request to the handler.

I believe it's actually failing b/c the `Form` data guard only supports `multipart/form-data` and `application/x-www-form-urlencoded`. If you cannot use one of these content types, you will need to use an external crate.

> can you recommend any 3rd-party crate(s) that do this and work well w/ Rocket?

Not right now. It's a logical enhancement for my `rocket-multipart` crate though, so I'll look into adding the capability to parse `multipart/mixed` content.

**Reply by raif-s-naffah** [reply]
thanks Matthew for your prompt response.  much appreciated!

FYI when implementing the parsing functionality in `rocket-multipart`, the standard in question also relies (or expects) incoming Parts to each expose their Headers so the implementation can check for the presence + contents of the following ones:

* content-type,
* content-length,
* custom headers --those named `x-<something>`.

may be something like [`multer`'s `Field`][1] would be worth considering.

[1]: https://docs.rs/multer/3.1.0/multer/struct.Field.html#method.headers

**Reply by the10thWiz** [reply, marked as answer]
Well, I've taken a [stab](https://crates.io/crates/rocket-multipart) at it (and released as v0.1.1). Let me know if there are any additional features that might be useful.

**Reply by raif-s-naffah** [reply]
thank you kindly Matthew!

i'll have a closer look at it tomorrow but quickly perusing the docs i noticed a method named [`contenty_type`][1].  was that intentional?

also, just to be sure, i should be able to wire this in my handler like so...
```rust
async fn put(mut data: MultipartReader<'_>) -> MyResponse { }
```
yes?

finally, if i find i have more questions, is it best to post them here or [there][2]?

thanks again + have a nice weekend.

[1]: https://docs.rs/rocket-multipart/0.1.1/rocket_multipart/struct.MultipartReader.html#method.contenty_type
[2]: https://github.com/the10thWiz/rocket-multipart/issues

**Reply by the10thWiz** [reply]
The content_type typo was not intentional. I'm going to fix that (I'll need to figure out what version I need to bump, but oh well).

> also, just to be sure, i should be able to wire this in my handler like so...

If you look at the source file in the repo, I have a route (for testing) that's pretty much what you describe.

It's probably better to post questions to my git repos.

**Reply by raif-s-naffah** [reply]
hi Matthew,

couldn't resist the temptation and quickly tested your latest release and it works!

i'm marking this question as answered + will post followup questions relevant to this aspect on the `rocket-multipart` project.

thanks again for your response!
